### PR TITLE
Decoupling engine backend part 1

### PR DESCRIPTION
**Pull request by Horneth**

### CODE REVIEW

File: core/src/main/scala/cromwell/core/package.scala [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package cromwell
+
+import wdl4s.values.{SymbolHash, WdlValue}
+
+import scalaz._
+
+/**
+  * Created by tjeandet on 3/21/16.
+  */
```
Comment (mcovarr) [inline core/src/main/scala/cromwell/core/package.scala]: nix

**Comment by Horneth** [comment]
This is less than I hoped to acheive, I tried to clean up the CallExecutionBackendJobOutputFInalDatabase-Key mayhem but gave up for now. If that's something that should be in this PR I can try again

File: engine/src/main/scala/cromwell/engine/backend/Backend.scala [hunk 2]
```diff
@@ -140,12 +147,10 @@ trait Backend {
 
   def executionInfoKeys: List[String]
 
-  def callRootPathWithBaseRoot(jobDescriptor: BackendCallJobDescriptor, baseRoot: String): Path = {
-    jobDescriptor.key.callRootPathWithBaseRoot(jobDescriptor.workflowDescriptor, baseRoot)
-  }
+  def callRootPathWithBaseRoot(key: CallKey, descriptor: WorkflowDescriptor, baseRoot: String): Path = Backend.callRootPathWithBaseRoot(key, descriptor, baseRoot)
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/backend/Backend.scala]: why is the key + workflow descriptor broken out from `JobDescriptor`?

Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/backend/Backend.scala]: This is a left over from my failed attempt to decouple the various keys, I'll revert it

File: engine/src/main/scala/cromwell/engine/backend/package.scala [hunk 3]
```diff
@@ -1,16 +1,30 @@
...
+  }
+
   // Ordered by shards, and then ordered by attempts
   type AttemptedCallLogs = Seq[Seq[CallLogs]]
   // Grouped by FQNS
   type WorkflowLogs = Map[FullyQualifiedName, AttemptedCallLogs]
 
   implicit class EnhancedFutureFuture[A](val ffa: Future[Future[A]])(implicit ec: ExecutionContext) {
-    def flatten: Future[A] = ffa flatMap { fa => fa }
+    def flatten: Future[A] = ffa flatMap identity
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/backend/package.scala]: :smile:

File: engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala [hunk 4]
```diff
@@ -117,7 +117,7 @@ object WorkflowActor {
       // This only does the initialization for a newly created workflow.  For a restarted workflow we should be able
       // to assume the adjusted symbols already exist in the DB, but is it safe to assume the staged files are in place?
       actor.initializeWorkflow match {
-        case Success(inputs) => actor.createWorkflow(inputs)
+        case Success(_) => actor.createWorkflow(actor.workflow.actualInputs)
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala]: why do we need to call a method on the actor parameterized by data on the actor?

File: core/src/main/scala/cromwell/core/package.scala [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+package object core {
+  type ErrorOr[+A] = ValidationNel[String, A]
+
+  type LocallyQualifiedName = String
+  case class CallOutput(wdlValue: WdlValue, hash: Option[SymbolHash])
+  type CallOutputs = Map[LocallyQualifiedName, CallOutput]
+
```
Comment (geoffjentry) [inline core/src/main/scala/cromwell/core/package.scala]: take this \n and stick it after the }, and then you'll have made both myself and khalid happy

File: engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala (hunk 4 above)
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala]: Good point
The change here comes form the fact that `initializeWorkflow` was returning "potentially" modified inputs but none of the backends were actually modifying them.
I'll clean that up

Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala]: Why is a method being directly called inside an actor in the first place?

File: engine/src/main/scala/cromwell/engine/workflow/ExecutionStoreKey.scala [hunk 6]
```diff
@@ -1,47 +1,32 @@
...
 import cromwell.engine.finalcall.FinalCall
-import wdl4s._
-import cromwell.engine.{WorkflowDescriptor, ExecutionStatus}
 import cromwell.engine.workflow.WorkflowActor.ExecutionStore
+import wdl4s._
 
 import scala.language.postfixOps
 
-sealed trait ExecutionStoreKey {
+sealed trait ExecutionStoreKey extends JobKey {
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/ExecutionStoreKey.scala]: Hmm this feels a little weird, the keys for the WA's internal execution store are extending a backend trait.  I probably shouldn't think about this too much.  :smile:

File: engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala (hunk 4 above)
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala]: IIRC it's part of the back and forth between the WA and the `StartMessage` subtype

File: project/Dependencies.scala [hunk 7]
```diff
@@ -54,9 +56,14 @@ object Dependencies {
...
 
+  val coreDependencies = List(
+    wdl4sDependency,
+    "com.typesafe" % "config" % "1.3.0"
+  )
+
   val engineDependencies = List(
     "org.broadinstitute" %% "lenthall" % lenthallV,
-    "org.broadinstitute" %% "wdl4s" % wdl4sV,
+    wdl4sDependency,
```
Comment (mcovarr) [inline project/Dependencies.scala]: is this necessary or doesn't the engine get core's dependencies transitively?

Comment (geoffjentry) [inline project/Dependencies.scala]: They inherit them but I've generally seen people list them out explicitly when I look at multi-project stuff so YMMV

File: engine/src/main/scala/cromwell/engine/workflow/ExecutionStoreKey.scala (hunk 6 above)
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/workflow/ExecutionStoreKey.scala]: I think I managed to convince myself it was ok, saying that Backend can define a trait containing the stuff it needs to do its thing, and Engine can build on top of that to add its own internal sauce as it pleases..

File: engine/src/test/scala/cromwell/engine/workflow/BackendCallKeySpec.scala [hunk 8]
```diff
@@ -18,15 +18,15 @@ class BackendCallKeySpec extends FlatSpec with Matchers with Mockito {
...
 
-    new BackendCallKey(call, None, 1).callRootPathWithBaseRoot(wd, root).toString shouldBe "/root/wfName/uuid/call-callName"
-    new BackendCallKey(call, Some(0), 1).callRootPathWithBaseRoot(wd, root).toString shouldBe "/root/wfName/uuid/call-callName/shard-0"
-    new BackendCallKey(call, Some(0), 2).callRootPathWithBaseRoot(wd, root).toString shouldBe "/root/wfName/uuid/call-callName/shard-0/attempt-2"
-    new BackendCallKey(call, None, 2).callRootPathWithBaseRoot(wd, root).toString shouldBe "/root/wfName/uuid/call-callName/attempt-2"
+    Backend.callRootPathWithBaseRoot(new BackendCallKey(call, None, 1), wd, root).toString shouldBe "/root/wfName/uuid/call-callName"
+    Backend.callRootPathWithBaseRoot(new BackendCallKey(call, Some(0), 1), wd, root).toString shouldBe "/root/wfName/uuid/call-callName/shard-0"
+    Backend.callRootPathWithBaseRoot(new BackendCallKey(call, Some(0), 2), wd, root).toString shouldBe "/root/wfName/uuid/call-callName/shard-0/attempt-2"
+    Backend.callRootPathWithBaseRoot(new BackendCallKey(call, None, 2), wd, root).toString shouldBe "/root/wfName/uuid/call-callName/attempt-2"
```
Comment (mcovarr) [inline engine/src/test/scala/cromwell/engine/workflow/BackendCallKeySpec.scala]: not sure I follow the reason for this change

File: engine/src/main/scala/cromwell/engine/workflow/ExecutionStoreKey.scala (hunk 6 above)
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/ExecutionStoreKey.scala]: ok

**Comment by mcovarr** [comment]
currently not compiling

File: engine/src/test/scala/cromwell/engine/workflow/BackendCallKeySpec.scala (hunk 8 above)
Comment (Horneth) [inline engine/src/test/scala/cromwell/engine/workflow/BackendCallKeySpec.scala]: I moved the code that creates the path for a call directory from the key to the Backend. I remember people being surprised that this was in the key class. It was also probably trying to remove a dependency even though I admit I don't remember which one specifically -_-

File: engine/src/main/scala/cromwell/engine/backend/jes/JesBackend.scala [hunk 9]
```diff
@@ -72,16 +72,10 @@ object JesBackend {
...
 
-  private val BuildJobKey: (Execution, Seq[ExecutionInfo]) => JobKey = (e: Execution, eis: Seq[ExecutionInfo]) => {
-    JesJobKey(eis.find(_.key == JesBackend.InfoKeys.JesRunId).get.value.get)
-  }
-
-  // Decoration around WorkflowDescriptor to generate bucket names and the like
-  implicit class JesWorkflowDescriptor(val descriptor: WorkflowDescriptor)
-    extends JesBackend(descriptor.backend.actorSystem) {
-    def callDir(key: BackendCallKey) = key.callRootPathWithBaseRoot(descriptor, rootPath(descriptor.workflowOptions))
+  private val BuildJobKey: (Execution, Seq[ExecutionInfo]) => BackendJobKey = (e: Execution, eis: Seq[ExecutionInfo]) => {
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: `BuildBackendJobKey`?

**Comment by mcovarr** [comment]
:+1: thanks for taking this on, I know it's super-boring but this is getting us a lot closer to where we need to be :tada:

File: engine/src/main/scala/cromwell/util/TryUtil.scala [hunk 10]
```diff
@@ -4,13 +4,14 @@ import java.io.{PrintWriter, StringWriter}
...
 import cromwell.logging.WorkflowLogger
+import org.apache.commons.lang3.exception.ExceptionUtils
 
 import scala.language.postfixOps
 import scala.util.{Failure, Success, Try}
 
 case class AggregatedException(exceptions: Seq[Throwable], prefixError: String = "") extends Exception {
   override def getMessage: String = {
-    prefixError + exceptions.map(_.getMessage).mkString("\n")
+    prefixError + exceptions.map(e => s"${e.getMessage}\n${ExceptionUtils.getStackTrace(e)}").mkString("\n")
```
Comment (kshakir) [inline engine/src/main/scala/cromwell/util/TryUtil.scala]: Moving the stack traces into the message… For some reason this horrifies and impresses me at the same time.

File: engine/src/main/scala/cromwell/webservice/WorkflowQueryKey.scala [hunk 11]
```diff
@@ -1,12 +1,13 @@
...
-import cromwell.engine.{ErrorOr, WorkflowState}
+import cromwell.core.ErrorOr
+import cromwell.engine.WorkflowState
 import org.joda.time.DateTime
 
 import scala.language.{postfixOps, reflectiveCalls}
 import scala.util.{Success, Try}
 import scalaz.Scalaz._
-import scalaz.{Success => _, _}
+import scalaz.{Success => _}
```
Comment (kshakir) [inline engine/src/main/scala/cromwell/webservice/WorkflowQueryKey.scala]: Do you still need this line?

File: engine/src/main/scala/cromwell/engine/callexecution/CallExecutionActor.scala [hunk 12]
```diff
@@ -3,7 +3,7 @@ package cromwell.engine.callexecution
 import akka.actor.{Actor, Props}
 import akka.event.{Logging, LoggingReceive}
 import com.google.api.client.util.ExponentialBackOff
-import cromwell.engine.backend.{ExecutionHandle, JobKey, _}
+import cromwell.engine.backend.{ExecutionHandle, BackendJobKey, _}
```
Comment (kshakir) [inline engine/src/main/scala/cromwell/engine/callexecution/CallExecutionActor.scala]: Can this be `import cromwell.engine.backend._`?

File: engine/src/test/scala/cromwell/engine/WorkflowDescriptorSpec.scala [hunk 13]
```diff
@@ -5,8 +5,9 @@ import java.nio.file.{Files, Paths}
 import better.files._
 import com.typesafe.config.ConfigFactory
 import cromwell.CromwellTestkitSpec
+import cromwell.core.WorkflowId
 import cromwell.engine.backend.io._
-import cromwell.engine.backend.CallMetadata
+import cromwell.engine.backend.{WorkflowDescriptor, CallMetadata}
```
Comment (kshakir) [inline engine/src/test/scala/cromwell/engine/WorkflowDescriptorSpec.scala]: Anytime a `main` class moves, its `test` should move too.

For example: once this spec is in the right package, this import will not be necessary.

**Comment by kshakir** [comment]
Minor updates, then :+1:

File: engine/src/main/scala/cromwell/util/TryUtil.scala (hunk 10 above)
Comment (Horneth) [inline engine/src/main/scala/cromwell/util/TryUtil.scala]: I know.. The problem with this AggregatedException is that is concatenates the messages but you lose all the stacktraces. I know it's ugly but it makes it hard sometimes to know what happened when postProcessing fails for example. We just get these "500 Internal Server Error" messages with no real way to know which part of the processing failed.
If you have a better idea I'm all ears ! :)

Comment (mcovarr) [inline engine/src/main/scala/cromwell/util/TryUtil.scala]: maybe log the constituent exceptions here?

File: engine/src/main/scala/cromwell/engine/backend/Backend.scala [hunk 14]
```diff
@@ -50,9 +47,21 @@ object Backend {
...
+  private val CallPrefix = "call"
+  private val ShardPrefix = "shard"
+  private val AttemptPrefix = "attempt"
+
+  def callRootPathWithBaseRoot(jobDescriptor: BackendCallJobDescriptor, baseRoot: String): Path = {
+    import io._
+    val call = s"$CallPrefix-${jobDescriptor.key.scope.unqualifiedName}"
+    val shard = jobDescriptor.key.index map { s => s"$ShardPrefix-$s" } getOrElse ""
+    val retry = if (jobDescriptor.key.attempt > 1) s"$AttemptPrefix-${jobDescriptor.key.attempt}" else ""
+    jobDescriptor.workflowDescriptor.workflowRootPathWithBaseRoot(baseRoot).resolve(call).resolve(shard).resolve(retry).asDirectory
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/backend/Backend.scala]: almost long enough to warrant a `foldLeft` over the path components... :smile:

File: engine/src/main/scala/cromwell/engine/backend/jes/JesBackend.scala [hunk 15]
```diff
@@ -252,21 +246,17 @@ case class JesBackend(actorSystem: ActorSystem)
...
-    generateAuthJson(dockerConf, getGcsAuthInformation(workflow)) foreach { content =>
+    generateAuthJson(dockerConf, getGcsAuthInformation(workflow)) map { content =>
 
       val path = gcsAuthFilePath(workflow)
       def upload(prev: Option[Unit]): Unit = path.writeAsJson(content)
 
       log.info(s"Creating authentication file for workflow ${workflow.id} at \n ${path.toString}")
       withRetry(upload, log, s"Exception occurred while uploading auth file to $path")
-    }
+    } getOrElse Success(Unit)
```
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: Why not `Success(())`?

Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: Why is it better ?

File: engine/src/main/scala/cromwell/util/TryUtil.scala (hunk 10 above)
Comment (Horneth) [inline engine/src/main/scala/cromwell/util/TryUtil.scala]: Good idea it's probably cleaner

File: engine/src/main/scala/cromwell/engine/backend/jes/JesBackend.scala (hunk 15 above)
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: Technically you can stick anything in there but `()` is the idiomatic way of returning a unit

Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: Ok I was doing `()` before but I've seen `Unit` and thought it was better. I'll put `()` I really don't care either way

Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: TBH neither do I, it just jumped out at me when I was scrolling past

**Pull request merged**